daemon/k8s_watcher.go

```diff
@@ -1550,26 +1623,88 @@ func (d *Daemon) addCiliumNetworkPolicyV2(ciliumV2Store cache.Store, cnp *cilium
-			LastUpdated: cilium_v2.NewTimestamp(),
-		}
 		scopedLog.WithError(err).Warn("Unable to add CiliumNetworkPolicy")
 	} else {
-		cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
-			OK:          true,
-			LastUpdated: cilium_v2.NewTimestamp(),
-		}
 		scopedLog.Info("Imported CiliumNetworkPolicy")
 	}
 
-	go func() {
-		k8s.UpdateCNPStatusV2(ciliumNPClient.CiliumV2(), ciliumV2Store,
-			k8s.BackOffLoopTimeout, node.GetName(), cnp, cnpns)
-	}()
+	ctrlName := cnp.GetControllerName()
+	cnpCM.UpdateController(ctrlName,
+		controller.ControllerParams{
+			DoFunc: func() error {
+				cnpns := cilium_v2.CiliumNetworkPolicyNodeStatus{
+					OK: true,
+				}
+
+				d.policy.Mutex.RLock()
+				rev := d.policy.GetRevision()
+				d.policy.Mutex.RUnlock()
+
+				cnpns.Revision = rev
+
+				ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+				defer cancel()
+
+				waitForEPsErr := endpointmanager.WaitForEndpointsAtPolicyRev(ctx, rev)
+
+				// If the deadline was reached then not all endpoints are enforcing
+				// the given policy.
+				cnpns.Enforcing = waitForEPsErr == nil
+
+				serverRuleStore, exists, err := ciliumV2Store.Get(cnp)
+				if !exists {
+					// If the CNP doesn't exist we clean up the controller
+					cnpCM.RemoveController(ctrlName)
+					return nil
+				}
+				if err != nil {
+					scopedLog.WithError(err).Warn("Unable to find v2.CiliumNetworkPolicy in local cache")
+					// Should we remove the controller in this case?
+					return err
+				}
+
+				serverRule, ok := serverRuleStore.(*cilium_v2.CiliumNetworkPolicy)
+				if !ok {
+					scopedLog.WithFields(logrus.Fields{
+						logfields.CiliumNetworkPolicy: logfields.Repr(serverRuleStore),
+					}).Warn("Received object of unknown type from API server, expecting v2.CiliumNetworkPolicy")
+					return errors.New("Received object of unknown type from API server, expecting v2.CiliumNetworkPolicy")
+				}
+
+				serverRuleCpy := serverRule.DeepCopy()
+				_, err2 := serverRuleCpy.Parse()
+				if err2 != nil {
+					// If we can't parse the rule then we should signalize
+					// it in the status
+					log.WithError(err2).WithField(logfields.Object, logfields.Repr(serverRuleCpy)).
+						Warn("Error parsing new CiliumNetworkPolicy rule")
+				}
+				if err != nil {
+					cnpns.Error = err.Error()
+					cnpns.OK = false
+				} else if err2 != nil {
+					cnpns.Error = err2.Error()
+					cnpns.OK = false
+				}
+
+				nodeName := node.GetName()
+
+				cnpns.LastUpdated = cilium_v2.NewTimestamp()
+				serverRuleCpy.SetPolicyStatus(nodeName, cnpns)
+				ns := cilium_io.ExtractNamespace(&serverRuleCpy.ObjectMeta)
+
+				_, err2 = ciliumNPClient.CiliumV2().CiliumNetworkPolicies(ns).Update(serverRuleCpy)
+				if err2 == nil {
+					scopedLog.WithField("status", serverRuleCpy.Status).Debug("successfully updated with status")
+				} else {
+					return err2
+				}
+				return waitForEPsErr
+			},
+			RunInterval: time.Duration(1) * time.Minute,
```

Comment:
This seems too low. The controller gets kicked off on policy change which should be primary driver. The RunInterval should only be there to annotate the CRD if it got messed up. I would default to 5 minutes. We might even need an interval that gets bigger depending on the number of nodes.

Reply:
We can simply remove the `RunInterval` since the CNP are re-fetched every 5 minutes. (When a refresh happens the controller is deleted and re-added)

Reply:
Don't we diff and avoid the policy change if nothing changed?